pkg/gadgets/trace/tcpdump/tracer/tracer.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracer
+
+import (
+	"fmt"
+	"syscall"
+	"unsafe"
+
+	"github.com/google/gopacket/layers"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/tcpdump/types"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/rawsock"
+	eventtypes "github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+	"golang.org/x/net/bpf"
+	"golang.org/x/sys/unix"
+)
+
+type Config struct {
+	FilterString string
+	SnapLen      int
+}
+
+type link struct {
+	sockFd int
+
+	// users count how many users called Attach(). This can happen for two reasons:
+	// 1. several containers in a pod (sharing the netns)
+	// 2. pods with networkHost=true
+	users int
+}
+
+type Tracer struct {
+	program []bpf.RawInstruction
+
+	// key: namespace/podname
+	// value: Tracelet
+	attachments map[string]*link
+}
+
+func NewTracer(config *Config) (*Tracer, error) {
+	t := &Tracer{
+		attachments: map[string]*link{},
+	}
+
+	ins, err := TcpdumpExprToBPF(config.FilterString, layers.LinkTypeEthernet, config.SnapLen)
```

Comment:
Perhaps `TcpdumpExprToBPF()` could return `[]bpf.RawInstruction` directly to avoid the `Disassemble()`/`Assemble()` calls?